### PR TITLE
Prevent over eager broadcast cooldown

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

- Allow a few attempts (configurable) to broadcast before entering a
  cooldown period for broadcasts (configurable) per peer
- ~~`PeerConnectionStats` replaced with `LastConnectionState`.
  `LastConnectionState` specifically deals with the last connection~~
- `PeerConnectionStats` includes details on the last connection  attempt 
and how many times a connection failed in a row.
- Test for select_peers in DHT
- Should fix flakiness for simultaneous connect test (e.g. https://app.circleci.com/jobs/github/tari-project/tari/7232)

## Motivation and Context

When using a 1:many broadcast strategy, peers which are not contactable should be temporarily excluded. This allows broadcasting to 'expand' (in terms of the distance metric) if a number of neighbouring nodes go offline. An attempt to contact nodes directly is still always attempted.

## Types of changes

* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/peer_manager/mod.rs [hunk 1]
```diff
@@ -40,6 +40,10 @@
 //! # use std::sync::Arc;
 //! # use tari_storage::LMDBWrapper;
 //!
+//! # // Remove database files in case they aren't compatible
```
Comment (CjS77) [inline comms/src/peer_manager/mod.rs]: Probably just `norun` this test.. I suspect writing to the FS in doc tests will be frowned upon in e.g. docs.rs and other places

File: comms/src/peer_manager/peer.rs [hunk 2]
```diff
@@ -39,13 +43,76 @@ bitflags! {
...
+    pub fn failed_at(&self) -> Option<&NaiveDateTime> {
+        match self {
+            LastConnectionState::Failed { failed_at, .. } => Some(failed_at),
+            _ => None,
+        }
+    }
+
+    /// Returns the Duration since the last failed connection occurred. None is returned if the `LastConnectionState` is
+    /// not `Failed`
+    pub fn failed_since_now(&self) -> Option<Duration> {
```
Comment (CjS77) [inline comms/src/peer_manager/peer.rs]: This is a weird name. time_since_last_failure?

File: comms/src/peer_manager/mod.rs (hunk 1 above)
Comment (sdbondi) [reply comms/src/peer_manager/mod.rs]: Agreed

File: comms/src/peer_manager/peer.rs (hunk 2 above)
Comment (sdbondi) [reply comms/src/peer_manager/peer.rs]: One day I'll figure out this naming thing 😆

**Pull request merged**